docs/proposals/petset.md

```diff
@@ -0,0 +1,533 @@
+  than running a clustered system off Kube. Authors should be given tools to guard against common cluster
+  failure modes (split brain, phantom member) to prevent introducing more failure modes. Sophisticated
+  distributed systems designers can implement more sophisticated solutions than PetSet if necessary -
+  new users should not become vulnerable to additional failure modes through an overly flexible design.
+* **Limited scaling** - While flexible scaling is important for some clusters, other examples of clusters
+  do not change scale without significant external intervention. Human intervention may be required after
+  scaling. Changing scale during cluster operation can lead to split brain in quorum systems. It should be
+  possible to scale easily, but the details of making that safe belong to the pods and image authors.
+* **No generic cluster lifecycle** - Rather than design a general purpose lifecycle for clustered software,
+  focus on ensuring the information necessary for the software to function is available. For example,
+  rather than providing a "post-creation" hook invoked when the cluster is complete, provide the necessary
+  information to the "first" (or last) pod to determine the identity of the remaining cluster members and
+  allow it to manage its own initialization.
+* **External access direct to cluster members is out of scope** - exposing pods to consumers that cannot
+  access the pod network is out of scope, because we do not currently support headless services being
+  exposed via NodePort. A workaround is to allow external clients to access the pod network, or to
+  create one NodePort service per member. A future design should cover headless external service access.
+
+
+## Proposed Design
+
+Add a new resource to Kubernetes to represent a set of pods that are individually distinct but each
+individual can safely be replaced-- the name **PetSet** (working name) is chosen to convey that the
+individual members of the set are themselves "pets" and thus each one is important. A relevant analogy
+is that a PetSet is composed of pets, but the pets are like goldfish. If you have a blue, red, and
+yellow goldfish, and the red goldfish dies, you replace it with another red goldfish and no one would
+notice. If you suddenly have three red goldfish, someone will notice.
+
+The PetSet is responsible for creating and maintaining a set of **identities** and ensuring that there is
+one pod and zero or more **supporting resources** for each identity. There should never be more than one pod
+or unique supporting resource per identity at any one time. A new pod can be created for an identity only
+if a previous pod has been fully terminated (reached its graceful termination limit or cleanly exited).
+
+```
+type PetSet struct {
+    ObjectMeta
+
+    Spec PetSetSpec
+    ...
+}
+
+type PetSetSpec struct {
+    // Number of instances in the set
+    Replicas *int
```

Comment:
I might consider calling this "Members", "Instances", or something else other than "Replicas", since "Replicas" has an implication of fungibility in my mind, though I realize introducing a different term may also confuse some people.